Include/internal/pycore_lock.h

```diff
@@ -0,0 +1,161 @@
+// Lightweight locks and other synchronization mechanisms.
+//
+// These implementations are based on WebKit's WTF::Lock. See
+// https://webkit.org/blog/6161/locking-in-webkit/ for a description of the
+// design.
+#ifndef Py_INTERNAL_LOCK_H
+#define Py_INTERNAL_LOCK_H
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#ifndef Py_BUILD_CORE
+#  error "this header requires Py_BUILD_CORE define"
+#endif
+
+#include "pycore_time.h"          // _PyTime_t
+
+
+// A mutex that occupies one byte. The lock can be zero initialized.
+//
+// Only the two least significant bits are used. The remaining bits should be
+// zero:
+// 0b00: unlocked
+// 0b01: locked
+// 0b10: unlocked and has parked threads
+// 0b11: locked and has parked threads
+//
+// Typical initialization:
+//   PyMutex m;
+//   memset(&m, 0, sizeof(m));
+//
+// Typical usage:
+//   PyMutex_Lock(&m);
+//   ...
+//   PyMutex_Unlock(&m);
+typedef struct _PyMutex {
+    uint8_t v;
```

Comment:
Does this need to be byte-aligned when embedded?

Reply:
Also, FWIW, I've come to appreciate a pattern we use elsewhere for bit fields in a struct like this.  Here it would look something like this:

```suggestion
    struct {
        unsigned int locked:1;
        unsigned int has_parked:1;
        /* padding to align to 1 byte */
        unsigned int :6;
    } v;
```

Would that be less efficient or compile differently?

Would it be too clunky (e.g. due to casts to `uint8_t`)?  (I suppose we could use a union to expose it as `uint8_t`.  Is that getting too clever?)

Reply:
It doesn't need any extra care regarding alignment. (Minimum alignment in C is 1 byte -- you can't address or align something smaller than that.)

Using a bitfield ends up being too clunky, because we need to use atomic operations.
- We can't do atomic operations on `struct` types or bitfields, just scalar types because we need to wrap MSVC atomic intrinsics as well as C11 style atomics. 
- The suggested definition is 4 bytes (not 1 byte!) on [GCC, Clang, and MSVC](https://gcc.godbolt.org/z/MfzP6xP7d). It's tricky to get the exact size you want with bitfields and in this case it's important that the mutex is only a single byte.

Reply:
Thanks for the explanation.  That makes sense.